File: php/testcode/fileupload_029.php
<?php
require_once __DIR__ . '/shared.php';

// vuln-code-snippet start php_upload_guessable_filename
function fileupload029(BenchmarkRequest $req): BenchmarkResponse {
    $file = $_FILES['f'];
    $dir  = '/var/www/uploads/';
    move_uploaded_file($file['tmp_name'], $dir . $file['name']); // vuln-code-snippet vuln-line php_upload_guessable_filename
    return BenchmarkResponse::ok('Uploaded as ' . $file['name']);
}
// vuln-code-snippet end php_upload_guessable_filename

File: rust/testcode/fileupload_029.rs
//! CWE-434: Upload path constructed from user-supplied directory and filename without validation.

// vuln-code-snippet start testcodeFileupload029
pub fn handle(req: &super::shared::BenchmarkRequest) -> super::shared::BenchmarkResponse {
    let dir = req.param("dir");
    let name = req.param("name");
    let data = req.param("data");
    let dest = build_path(&dir, &name);
    let _ = std::fs::write(&dest, data.as_bytes()); // vuln-code-snippet target-line testcodeFileupload029
    super::shared::BenchmarkResponse::ok("Uploaded")
}

fn build_path(dir: &str, name: &str) -> String {
    format!("{}/{}", dir, name)
}
// vuln-code-snippet end testcodeFileupload029
